api/odilo.py

```diff
@@ -0,0 +1,495 @@
+            OdiloBibliographicCoverageProvider(
+                collection, api_class=self
+            )
+        )
+
+    def patron_request(self, patron, pin, url, extra_headers={}, data=None, exception_on_401=False, method=None):
+        """Make an HTTP request on behalf of a patron.
+
+        The results are never cached.
+        """
+        headers = dict(Authorization="Bearer %s" % self.token)
+        headers['Content-Type'] = 'application/json'
+        headers.update(extra_headers)
+
+        if method and method.lower() in ('get', 'post', 'put', 'delete'):
+            method = method.lower()
+        else:
+            if data:
+                method = 'post'
+            else:
+                method = 'get'
+
+        response = HTTP.request_with_timeout(method, self.library_api_base_url + url, headers=headers, data=data,
+                                             timeout=60)
+        if response.status_code == 401:
+            if exception_on_401:
+                # This is our second try. Give up.
+                raise Exception("Something's wrong with the patron OAuth Bearer Token!")
+            else:
+                # Refresh the token and try again.
+                self.check_creds(True)
+                return self.patron_request(patron, pin, url, extra_headers, data, True)
+        else:
+            return response
+
+    def get_patron_credential(self, patron, pin):
+        """Create an OAuth token for the given patron."""
+
+        def refresh(credential):
+            return self.get_patron_access_token(credential, patron, pin)
+
+        return Credential.lookup(self._db, DataSource.ODILO, "OAuth Token", patron, refresh)
+
+    def get_patron_access_token(self, credential, patron, pin):
+        """Request an OAuth bearer token that allows us to act on
+        behalf of a specific patron.
+        """
+
+        self.client_key = patron
+        self.client_secret = pin
+        message = self.refresh_creds(credential)
+
+        if 'OK' == message:
+            return credential
+        else:
+            raise PatronAuthorizationFailedException(message)
+
+    def checkout(self, patron, pin, licensepool, internal_format):
+        """Check out a book on behalf of a patron.
+
+        :param patron: a Patron object for the patron who wants
+        to check out the book.
+
+        :param pin: The patron's alleged password.
+
+        :param licensepool: Identifier of the book to be checked out is 
+        attached to this licensepool.
+
+        :param internal_format: Represents the patron's desired book format.
+
+        :return: a LoanInfo object.
+        """
+        record_id = licensepool.identifier.identifier
+
+        # We have 2 formats, 'EBOOK_STREAMING' could be in our API ('PDF', 'EPUB'):
+        ebook_streaming = False
+        if internal_format == 'EBOOK_STREAMING':
+            ebook_streaming = True
+            internal_format = 'EPUB'  # Try first ebook_streaming as EPUB
```

Comment:
I'm not happy about this but I don't see a good solution to it except for storing custom media types in the delivery mechanisms. 

Can you change the Odilo API to understand 'EBOOK_STREAMING'? If it really doesn't matter to the client whether the underlying format is PDF or EPUB, there's no reason the client should have to keep track of the distinction and repeat it back to the server.